From 909e473fcdcab65177e95c5562589c7ae24d3f36 Mon Sep 17 00:00:00 2001
From: danascape <danascape@gmail.com>
Date: Sun, 19 Feb 2023 03:28:28 +0530
Subject: [PATCH] drivers: Add a simple echo module

* This was performed as a task for  Linux kernel Bug Fixing 2023
  (Spring).
* The module takes in a string as a parameter/argument and prints the
  same.

Test: insmod echo.ko string="saalim"

Output:
    [ 1948.620272] echo: Initialise...
    [ 1948.620273] The provided string: Saalim

Signed-off-by: danascape <danascape@gmail.com>
---
 Makefile |  7 +++++++
 echo.c   | 32 ++++++++++++++++++++++++++++++++
 2 files changed, 39 insertions(+)
 create mode 100644 Makefile
 create mode 100644 echo.c

diff --git a/Makefile b/Makefile
new file mode 100644
index 0000000..809936c
--- /dev/null
+++ b/Makefile
@@ -0,0 +1,7 @@
+obj-m += echo.o
+
+all:
+	make -C /lib/modules/$(shell uname -r)/build M=$(PWD) modules
+
+clean:
+	make -C /lib/modules/$(shell uname -r)/build M=$(PWD) clean
diff --git a/echo.c b/echo.c
new file mode 100644
index 0000000..d830579
--- /dev/null
+++ b/echo.c
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: GPL-2.0
+#include<linux/module.h>
+#include<linux/moduleparam.h>
+#include<linux/init.h>
+
+/*
+ * A simple echo module
+ */
+
+static char *string = "Default";
+
+module_param(string, charp, 0000);
+MODULE_PARM_DESC(string, "A character string");
+
+static int __init echo_init(void)
+{
+	pr_info("echo: Initialise...\n");
+	pr_info("The provided string: %s\n", string);
+	return 0;
+}
+
+static void __exit echo_exit(void)
+{
+	pr_info("echo: Exit...\n");
+}
+
+module_init(echo_init);
+module_exit(echo_exit);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Saalim Quadri <danascape@gmail.com>");
+MODULE_DESCRIPTION("Simple echo Module");
-- 
2.39.2

